osidb/tests/test_cvss.py

```diff
@@ -0,0 +1,127 @@
+
+        affect_cvss2 = AffectCVSSFactory(
+            affect=affect, version=CVSS.CVSSVersion.VERSION2
+        )
+        affect_cvss3 = AffectCVSSFactory(
+            affect=affect,
+            version=CVSS.CVSSVersion.VERSION3,
+            issuer=CVSS.CVSSIssuer.NIST,
+        )
+
+        all_flaw_cvss_scores = flaw.cvss_scores.all()
+        assert len(all_flaw_cvss_scores) == 2
+        assert flaw_cvss2 in all_flaw_cvss_scores
+        assert flaw_cvss3 in all_flaw_cvss_scores
+
+        all_affect_cvss_scores = affect.cvss_scores.all()
+        assert len(all_affect_cvss_scores) == 2
+        assert affect_cvss2 in all_affect_cvss_scores
+        assert affect_cvss3 in all_affect_cvss_scores
+
+        # Test flaw/affect-version-issuer uniqueness constraint.
+        with pytest.raises(
+            ValidationError,
+            match="Flaw cvss with this Flaw, Version and Issuer already exists.",
+        ):
+            duplicite_flaw_cvss2 = FlawCVSSFactory.build(
+                flaw=flaw,
+                version=CVSS.CVSSVersion.VERSION2,
+                issuer=CVSS.CVSSIssuer.REDHAT,
+            )
+            duplicite_flaw_cvss2.save()
+
+        with pytest.raises(
+            ValidationError,
+            match="Affect cvss with this Affect, Version and Issuer already exists.",
+        ):
+            duplicite_affect_cvss3 = AffectCVSSFactory.build(
+                affect=affect,
+                version=CVSS.CVSSVersion.VERSION3,
+                issuer=CVSS.CVSSIssuer.NIST,
+            )
+            duplicite_affect_cvss3.save()
+
+    @pytest.mark.parametrize(
+        "cvss3_vector",
+        [
+            "test",
+            "CVSS:3.1/AV:Z/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:N",  # wrong metric value Z
+            "AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:N",  # missing header
+            "AV:N/AC:L/Au:N/C:N/I:N/A:C",  # CVSSv2
+        ],
+    )
+    def test_validate_cvss3_string(self, cvss3_vector):
+        flaw = FlawFactory()
+        with pytest.raises(ValidationError, match="Invalid CVSS"):
+            FlawCVSSFactory(
+                flaw=flaw,
+                version=CVSS.CVSSVersion.VERSION3,
+                vector=cvss3_vector,
+            )
+
+    @pytest.mark.parametrize(
+        "cvss2_vector",
+        [
+            "test",
+            "AV:Z/AC:L/Au:N/C:N/I:N/A:C",  # wrong metric value Z
+            "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:N",  # CVSSv3
+        ],
+    )
+    def test_validate_cvss2_string(self, cvss2_vector):
+        with pytest.raises(ValidationError, match="Invalid CVSS"):
+            FlawCVSSFactory(version=CVSS.CVSSVersion.VERSION2, vector=cvss2_vector)
+
+    @pytest.mark.enable_signals
+    def test_cvss_recalculation(self):
+        cvss = FlawCVSSFactory(
+            vector="CVSS:3.1/AV:A/AC:H/PR:L/UI:R/S:C/C:L/I:L/A:L",
+            version=CVSS.CVSSVersion.VERSION3,
+        )
+        assert cvss.cvss_handle(cvss.vector).scores()[0] == cvss.score
```

Comment:
This kind of test is quite masturbatory and doesn't actually test anything because you're using untested code to do the testing, what if the logic of `cvss_handle` is wrong? The test will pass but it will be wrong.

I would rather do

```python
def test_cvss_recalculation(self, vector):
    cvss = FlawCVSSFactory(vector=vector, CVSS.CVSSVersion.VERSION3)
    assert cvss.score = CVSS3(vector).scores()[0]
```

Reply:
But does that make a difference if I use CVSS3 from cvsslib directly or via cvss_handle? :thinking:

Reply:
Imagine that in the future someone makes a change to the `CVSS_HANDLES` class variable and maps CVSSv3 to CVSS4 by mistake, your test will still pass (wrongfully) because it depends on the implementation of `cvss_handle`.

By being explicit and using `CVSS3` directly you no longer depend on the implementation of `cvss_handle` and check for what you actually want which is the calculated score being correct.

Reply:
I see. Makes sense, will change that.